proposals/public-suffix.md

```diff
@@ -0,0 +1,820 @@
+|--------------------|---------------------------------:|
+| allowIP            | IP Address (IPv4 of IPv6)        |
+| allowPlainSuffix   | is itself a known eTLD           |
+| allowUnknownSuffix | lacks a known eTLD               |
+
+The effect of each option when applied to an input hostname of the
+kind targetted by the option is to change the registrable domain
+from being `null` to being instead *the full input hostname itself*.
+
+###### Examples
+
+| Input hostname    | Option = true      | Registrable domain |
+|-------------------|--------------------|-------------------:|
+| 192.168.2.1       | allowIP            |        192.168.2.1 |
+| github.io         | allowPlainSuffix   |          github.io |
+| apple.pear.banana | allowUnknownSuffix |  apple.pear.banana |
+
+##### 2.2.2 Options: Justification
+
+Option `allowUnknownSuffix` supports use cases that need to target
+not only domains on the internet having known eTLDs, but also
+intranet hostnames having non-public (i.e. unknown) suffixes, or no suffix.
+
+Reviewers of this proposal note that if it were the case that non-domains
+were included by default, `getRegistrableDomain()` would effectively
+return a string for almost every input.
+
+As a result of the inclusion of unknown suffixes, the API implementation must
+take care to ensure input hostnames that are IP addresses (IPv4 or IPv6) are detected
+*before* running the algorithm. IP addresses must be treated differently,
+otherwise it would be possible for `100.200.30.2` and `100.200.31.2` to be
+mistakenly interpreted as belonging to the same organization (i.e. with unknown eTLD `.2`).
+
+Option `allowIP` supports use cases needing to obtain a hostname's site,
+which may be an IP address or a domain name.
+
+An example of such a use case is Firefox's [Search vs Navigate](#4-search-vs-navigate),
+which involves determining if an entry in the URL bar is a navigable site,
+or a search term. If this functionality was based purely on the return value
+of `getRegistrableDomain()`, i.e. navigate if nonnull or search if null,
+then IP addresses would incorrectly cause a search. By using the `allowIP` option,
+the return value for an input IP address would be the IP address itself instead of null,
+thereby causing the desired result of navigating instead of searching.
+
+Option `allowPlainSuffix` only exists because there are domains that do not have
+a registrable domain, due to themselves being PSL eTLDs, but can still be
+navigated to, such as github.io and blogspot.com.
+
+##### 2.2.3 Options: Discussion
+
+The effect of the options is that `getRegistrableDomain()` may return values
+that are not registrable domains in the strictest sense, e.g. they may
+be IP addresses.
+
+The author of this proposal is of the view that:
+
+1. Any method named `getXYZ()` should return a value of type `XYZ`. Therefore
+`getRegistrableDomain()` may not be the most suitable name, since it does
```

Comment:
I'm fine with either name - `publicSuffix.getRegistrableDomain()` or `publicSuffix.getDomain()`.

The current descriptions of the API schema frames the options as ways to change "registrable domain". I'd view it differently: independently of the method name, the definition of "registrable domain" is unchanged. The default return value is a registrable domain (as defined by URL spec). What the options do is to alter the return value to include more than just a registrable domain.

For clarity, it may make sense to change "Registrable domain" to "Returned domain" in the comparisons/tables, so that even if someone did not read the whole document, that they would not be confused when they see examples under such a column heading.

Reply:
Addressed in f1d81e5c10e93748e49c7a018ac9351b404fdd6d - changed the method names to the below as discussed, and reworded "registrable domain" to "returned domain" where appropriate:

```js
publicSuffix.getDomain()
publicSuffix.getKnownSuffix()
publicSuffix.isKnownSuffix()
```